File: python/backends/OpenVINO/yolo_openvino.py
'''
Author: taifyang  
Date: 2024-06-12 22:23:07
LastEditTime: 2025-12-23 08:26:56
Description: openvino inference class for YOLO algorithm
'''

import openvino as ov
from backends.yolo import *
from backends.utils import *

'''
description: openvino inference class for YOLO algorithm
'''
class YOLO_OpenVINO(YOLO):
    '''
    description:            construction method
    param {*} self          instance of class
    param {str} algo_type   algorithm type
    param {str} device_type device type
    param {str} model_type  model type
    param {str} model_path  model path
    return {*}
    '''      
    def __init__(self, algo_type:str, device_type:str, model_type:str, model_path:str) -> None:
        super().__init__()
        assert os.path.exists(model_path), 'model not exists!'
        assert device_type in ['CPU', 'GPU'], 'unsupported device type!'
        try:
            from openvino.runtime import Core
            core = Core()
        except:
            core = ov.Core()
        model  = core.read_model(model_path)
        self.algo_type = algo_type
        self.compiled_model = core.compile_model(model, device_name='GPU' if device_type=='GPU' else 'CPU')
        assert self.compiled_model, 'compile model failed!'
    
    '''
    description:    model inference
    param {*} self  instance of class
    return {*}
    '''       
    def process(self) -> None:
        self.outputs = self.compiled_model({0: self.inputs})

File: cpp/openvino/yolo_openvino.cpp
/* 
 * @Author: taifyang
 * @Date: 2024-06-12 09:26:41
 * @LastEditTime: 2025-12-21 22:41:31
 * @Description: openvino inference source file for YOLO algorithm
 */

#include "yolo_openvino.h"

void YOLO_OpenVINO::init(const Algo_Type algo_type, const Device_Type device_type, const Model_Type model_type, const std::string model_path)
{
	m_algo_type = algo_type;

	if(!std::filesystem::exists(model_path))
	{
		std::cerr << "model not exists!" << std::endl;
		std::exit(-1);
	}

	ov::Core core; //Initialize OpenVINO Runtime Core 
	ov::CompiledModel compiled_model;
	try
	{
		 compiled_model = core.compile_model(model_path, device_type == GPU ? "GPU" : "CPU"); //Compile the Model 
	}
	catch (const std::exception& e)
	{
		std::cerr << "openvino load model failed!" << std::endl;
		std::exit(-1);
	}

	m_infer_request = compiled_model.create_infer_request(); //Create an Inference Request 
	m_input_port = compiled_model.input(); //Get input port for model with one input
}
